### ISSUE TITLE
Error occurred when retrieving time series data

### ISSUE THREAD

**Issue by tharinduwijewardane** [Azure/azure-kusto-go#26]
Following error occurs when I execute a query with `make-series` clause.

```
Kind(KInternal): in row 0: column sum_value, conversion error: Column with type 'dynamic' was not stored as a string, was []interface {}
```

Example query:
```
 metrics 
| where uuid == 'obs111bd266-cc3d-434f-99f1-122bdb53e16d'
| where callerConnectorType == 'ballerina/http/Caller'
| where timeSeriesId == 'metric_requests_total'
| make-series sum(value) on timestamp in range(datetime(2020-05-05), datetime(2020-05-07), 2h) by httpStatusCodeGroup
```

As I understood its because the generated column `sum_value` is of `dynamic` type. But it should be supported by the SDK and I believe this is a bug.
Further, I am able to execute the same query using both REST API and Azure query UI without any issue.

**Comment by tharinduwijewardane** [comment]
If it may help, here is the code I ran,

```
func Temp() {
        authorizer := kusto.Authorization{
		Config: auth.NewClientCredentialsConfig(ClientId, ClientSecret, TenantID),
	}

	kustoClient, err := kusto.New(Endpoint, authorizer)
	if err != nil {
		return err
	}

	iter, err := kustoClient.Query(context.TODO(), "choreopocadx-db", kusto.NewStmt(
		`metrics 
	| where uuid == 'obs111bd266-cc3d-434f-99f1-122bdb53e16d'
	| where callerConnectorType == 'ballerina/http/Caller'
	| where timeSeriesId == 'metric_requests_total'
	| make-series sum(value) on timestamp in range(datetime(2020-05-05), datetime(2020-05-07), 2h) by httpStatusCodeGroup
	`))
	if err != nil {
		fmt.Println(err)
		return
	}
	defer iter.Stop()

	// .Do() will call the function for every row in the table.
	err = iter.Do(
		func(row *table.Row) error {
			fmt.Println(row)
			return nil
		},
	)
	if err != nil {
		fmt.Println(err)
	}
}
```

**Comment by element-of-surprise** [comment]
This indeed looks like a bug.  It is a conversion error from the REST JSON into the local type.
This must be something I broke in the last change where we updated how we do all the conversions.  I must have missed a test somewhere in that change, though its strange the etoe test passed.  

I'll patch this in a release and look back at the etoe, see why that didn't fail on us.  

There is one breaking change we want to do and it is going to be around how we handle Dynamic.  Right now we end up storing them as map[string]interface{}, but we are going to end up with just []byte and allow the user to do conversion.  This is going to give more freedom to the user that exists now for a minor cost on the client itself.

**Comment by tharinduwijewardane** [comment]
Thanks for the attention. When will you be able to fix this? My work depends on this, so my other option is to use the REST API and decode the response myself (which i will not enjoy doing)

**Comment by element-of-surprise** [comment]
I try for a 1 week SLA for bugs (this is not an official SLA, just mine for the Go SDK).  I will try to do this tomorrow, but we have companies holidays Thurs/Fri for my group, it could get delayed.  

I don't imagine it going past the SLA.  Just want to set an expectation so that you can decide what you need to do.

**Comment by element-of-surprise** [comment]
It seems that this is my misinterpretation of the dynamic type. There are several pages on types and apparently I missed these items on the dynamic type:

- "An array of dynamic values, holding zero or more values with zero-based indexing."
- "A property bag that maps unique string values to dynamic values. The property bag has zero or more such mappings (called "slots"), indexed by the unique string values. The slots are unordered."

I don't think I can make the existing Kusto.Dynamic support this in a way that will work.  The current map[string]interface{} conversion was a flawed approach on my part.  

The internal dynamics of conversion are complicated, but basically this isn't a simple patch.  I have a CL that I think will fix this in the works, but with a breaking change.

**Comment by tharinduwijewardane** [comment]
Thanks for the effort. Please do take your time and fix it properly. I am getting my work done through the REST API for now. I will visit Go SDK again when things are stable.

Let me interest you in something I came across when trying to parse the responses from the REST API. https://github.com/Jeffail/gabs is a really cool tool for parsing json data.

**Comment by element-of-surprise** [comment]
Its a cool tool.  The problem for us is it likely to be slow looking at its internals.  

Go's JSON decoder (which gabs uses) is fairly general and works well when data follows certain rules.  Performance wise and in other ways, it falls apart when dealing with JSON that doesn't conform to its model.  And the Go author's consider the JSON library as locked for features.  This is a well known problem.

Third party packages that use Go's standard JSON lib fall prey to the same traps.  RawMessage, as an example, has to make copies of the bytes in the message for safety, but this is a huge cost over a large stream.

There are third party packages that do JSON encoding, but either they don't support streaming correctly or they require a lot of generated code in order to work.  

We use a modified version of the stdlib that deals with our unique use case.  We want our allocs to be low for internal decoding.  We only want to alloc for values we pass to the user.  

In this particular bug we, we have two decodes for dynamic:
- We depack map[string]interface{}
- We depack string

What we needed was an extra:
- We depack []interface{}

Because I had missed the []interface{}, I exposed map[string]interface{}, which isn't going to suffice for that in any way that isn't absolutely hacky (like m["array"], yuck).

We are going to treat dynamic as we treat Decimal.  We are going to give you the raw data and let you decide how you want to unpack it.  So you should be getting []byte, in which you can covert as you see fit, including using gabs to do queries into the byte stream.

But again, thanks for brining the bug to our attention.  The preview release is mainly to find minor bugs that slipped through in testing.  Seems you found one!

**Comment by element-of-surprise** [comment]
I've updated dynamic at tip to handle what I think are all cases.  This converts Dynamic.Value into []byte.

We will be adding an integration test for this before we push this into a release.  

But you should be able to give this a try using the commit hash.

**Comment by element-of-surprise** [comment]
Wanted to check in and see if you've had an opportunity to try tip yet?

**Comment by element-of-surprise** [comment]
We've released the new version at 0.2.0 that should have taken care of this problem.  As we've received no response on this bug, I am closing this assuming that the problem is resolved.
